### PR TITLE
MISUV-5247 Add SE | Choose a reporting method

### PR DESCRIPTION
**Pull request by EdwardDennis**

### CODE REVIEW

File: app/controllers/incomeSources/add/BusinessReportingMethodController.scala [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        case err: InternalServerException =>
+          Logger("application").error(s"${if (isAgent) "[Agent]"}" + s"Error getting ITSA Status : $err")
+          errorHandler.showInternalServerError()
+      }
+
+
+    } else {
+      Future.successful(Ok(customNotFoundErrorView()(user, messages)))
+    } recover {
+      case ex: Exception =>
```
Comment (joesheridan) [inline app/controllers/incomeSources/add/BusinessReportingMethodController.scala:95]: Have we not already handled recovery elsewhere? ie is this block redundant?

File: app/controllers/incomeSources/add/BusinessReportingMethodController.scala [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    val errorHandler: ShowInternalServerError = if (isAgent) itvcErrorHandlerAgent else itvcErrorHandler
+    val redirectUrl: Call = if (isAgent) controllers.incomeSources.add.routes.BusinessAddedController.showAgent(id) else
+      controllers.incomeSources.add.routes.BusinessAddedController.show(id)
+
+    if (incomeSourcesEnabled) {
+      businessReportingMethodService.checkITSAStatusCurrentYear.flatMap {
+        case true =>
+          businessReportingMethodService.getBusinessReportingMethodDetails(id).map {
+            case Some(viewModel) =>
+              Ok(view(
```
Comment (joesheridan) [inline app/controllers/incomeSources/add/BusinessReportingMethodController.scala:70]: does the nav bar render on this page, I wonder?

Comment (EdwardDennis) [reply app/controllers/incomeSources/add/BusinessReportingMethodController.scala:70]: It does

File: app/controllers/incomeSources/add/BusinessReportingMethodController.scala [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    val incomeSourcesEnabled: Boolean = isEnabled(IncomeSources)
+    val postAction: Call = if (isAgent) controllers.incomeSources.add.routes.BusinessReportingMethodController.submitAgent(id) else
+      controllers.incomeSources.add.routes.BusinessReportingMethodController.submit(id)
+    val errorHandler: ShowInternalServerError = if (isAgent) itvcErrorHandlerAgent else itvcErrorHandler
+    val redirectUrl: Call = if (isAgent) controllers.incomeSources.add.routes.BusinessAddedController.showAgent(id) else
+      controllers.incomeSources.add.routes.BusinessAddedController.show(id)
+
+    if (incomeSourcesEnabled) {
+      businessReportingMethodService.checkITSAStatusCurrentYear.flatMap {
+        case true =>
```
Comment (joesheridan) [inline app/controllers/incomeSources/add/BusinessReportingMethodController.scala:67]: checkITSAStatusCurrentYear - its unclear from the name what the boolean result means.
may I suggest: isITSAStatusEligible and then the true/false values make more sense

File: app/controllers/incomeSources/add/BusinessReportingMethodController.scala [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+    val incomeSourcesEnabled: Boolean = isEnabled(IncomeSources)
+    val postAction: Call = if (isAgent) controllers.incomeSources.add.routes.BusinessReportingMethodController.submitAgent(id) else
+      controllers.incomeSources.add.routes.BusinessReportingMethodController.submit(id)
+    val errorHandler: ShowInternalServerError = if (isAgent) itvcErrorHandlerAgent else itvcErrorHandler
+    val redirectUrl: Call = if (isAgent) controllers.incomeSources.add.routes.BusinessAddedController.showAgent(id) else
+      controllers.incomeSources.add.routes.BusinessAddedController.show(id)
+
+    if (incomeSourcesEnabled) {
+      businessReportingMethodService.checkITSAStatusCurrentYear.flatMap {
```
Comment (joesheridan) [inline app/controllers/incomeSources/add/BusinessReportingMethodController.scala:66]: this should be in ITSAStatusService or similar
a service should own some data, so I'm doubtful as to the concept of a businessReportingMethodService

File: app/controllers/incomeSources/add/BusinessReportingMethodController.scala [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+    val submitUrl: Call = if (isAgent) controllers.incomeSources.add.routes.BusinessReportingMethodController.submitAgent(id) else
+      controllers.incomeSources.add.routes.BusinessReportingMethodController.submit(id)
+
+    if (incomeSourcesEnabled) {
+      AddBusinessReportingMethodForm.form.bindFromRequest().fold(
+        hasErrors => {
+          val updatedForm = AddBusinessReportingMethodForm.updateErrorMessagesWithValues(hasErrors)
+
+          if (updatedForm.hasErrors) {
+            businessReportingMethodService.getBusinessReportingMethodDetails(id).map {
```
Comment (joesheridan) [inline app/controllers/incomeSources/add/BusinessReportingMethodController.scala:115]: no need to get the 1171 data again, as we already have it

File: app/controllers/incomeSources/add/BusinessReportingMethodController.scala [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    val postAction: Call = if (isAgent) controllers.incomeSources.add.routes.BusinessReportingMethodController.submitAgent(id) else
+      controllers.incomeSources.add.routes.BusinessReportingMethodController.submit(id)
+    val errorHandler: ShowInternalServerError = if (isAgent) itvcErrorHandlerAgent else itvcErrorHandler
+    val redirectUrl: Call = if (isAgent) controllers.incomeSources.add.routes.BusinessAddedController.showAgent(id) else
+      controllers.incomeSources.add.routes.BusinessAddedController.show(id)
+
+    if (incomeSourcesEnabled) {
+      businessReportingMethodService.checkITSAStatusCurrentYear.flatMap {
+        case true =>
+          businessReportingMethodService.getBusinessReportingMethodDetails(id).map {
```
Comment (joesheridan) [inline app/controllers/incomeSources/add/BusinessReportingMethodController.scala:68]: no need to get the 1171 data again, as we already have it

File: app/controllers/incomeSources/add/BusinessReportingMethodController.scala [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+    val redirectUrl: Call = if (isAgent) routes.BusinessAddedController.showAgent(id) else routes.BusinessAddedController.show(id)
+    val submitUrl: Call = if (isAgent) controllers.incomeSources.add.routes.BusinessReportingMethodController.submitAgent(id) else
+      controllers.incomeSources.add.routes.BusinessReportingMethodController.submit(id)
+
+    if (incomeSourcesEnabled) {
+      AddBusinessReportingMethodForm.form.bindFromRequest().fold(
+        hasErrors => {
+          val updatedForm = AddBusinessReportingMethodForm.updateErrorMessagesWithValues(hasErrors)
+
+          if (updatedForm.hasErrors) {
```
Comment (joesheridan) [inline app/controllers/incomeSources/add/BusinessReportingMethodController.scala:114]: does the hasErrors function not already predicate that there are errors. In which case, is this check redundant?

File: app/controllers/incomeSources/add/BusinessReportingMethodController.scala [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+                errorHandler.showInternalServerError()
+              case None =>
+                Logger("application").info(s"${if (isAgent) "[Agent]"}" + s" Updating tax year specific reporting method not required.")
+                Redirect(redirectUrl)
+            }
+
+          }
+        },
+        valid => {
+          businessReportingMethodService.updateIncomeSourceTaxYearSpecific(user.nino, id, valid).map {
```
Comment (joesheridan) [inline app/controllers/incomeSources/add/BusinessReportingMethodController.scala:147]: duplicates code from above - but code above may not be needed as per other comment..

File: app/models/itsaStatus/ITSAStatusResponse.scala [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+package models.itsaStatus
+
+import play.api.libs.json.{Format, Json}
+
+sealed trait ITSAStatusResponse
+
+case class ITSAStatusResponseModel(taxYear: String,
+                                   itsaStatusDetails: Option[List[StatusDetail]] = None) extends ITSAStatusResponse
```
Comment (joesheridan) [inline app/models/itsaStatus/ITSAStatusResponse.scala:24]: can we have more than one status for a year?

File: app/services/BusinessReportingMethodService.scala [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+
+import javax.inject.{Inject, Singleton}
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class BusinessReportingMethodService @Inject()(incomeTaxViewChangeConnector: IncomeTaxViewChangeConnector,
+                                               dateService: DateService,
+                                               implicit val appConfig: FrontendAppConfig) extends FeatureSwitching {
+
+  def checkITSAStatusCurrentYear(implicit user: MtdItUser[_], hc: HeaderCarrier, ec: ExecutionContext): Future[Boolean] = {
```
Comment (joesheridan) [inline app/services/BusinessReportingMethodService.scala:42]: this should be in an ITSAStatus service instead

File: app/services/BusinessReportingMethodService.scala [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+      case Right(itsaStatus) =>
+        val isMandatedOrVoluntary = itsaStatus.exists(_.itsaStatusDetails.exists(_.exists(_.isMandatedOrVoluntary)))
+        Future.successful(isMandatedOrVoluntary)
+      case Left(x: ITSAStatusResponseError) =>
+        Logger("application").error(s"[BusinessReportingMethodService][checkITSAStatusCurrentYear] $x")
+        Future.failed(new InternalServerException("[BusinessReportingMethodService][checkITSAStatusCurrentYear] - Failed to retrieve ITSAStatus"))
+    }
+  }
+
+  def isTaxYearCrystallised(ty: Int)(implicit user: MtdItUser[_], hc: HeaderCarrier, ec: ExecutionContext): Future[Option[Boolean]] = {
```
Comment (joesheridan) [inline app/services/BusinessReportingMethodService.scala:60]: should be in a calculation service or crystallisationService?

File: app/services/BusinessReportingMethodService.scala [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+      incomeTaxViewChangeConnector.getCalculationList(Nino(user.nino), tyRange).flatMap {
+        case res: CalculationListModel => Future.successful(res.crystallised)
+        case err: CalculationListErrorModel => Future.failed(new InternalServerException(err.message))
+      }
+    }
+
+  }
+
+  def getBusinessReportingMethodDetails(incomeSourceId: String)(implicit user: MtdItUser[_], hc: HeaderCarrier, ec: ExecutionContext): Future[Option[BusinessReportingMethodViewModel]] = {
+    incomeTaxViewChangeConnector.getBusinessDetails(user.nino).flatMap {
```
Comment (joesheridan) [inline app/services/BusinessReportingMethodService.scala:77]: we shouldn't need to get the income source details again - we already do it... twice!!

File: app/services/BusinessReportingMethodService.scala [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+      case value: IncomeSourceDetailsModel =>
+        val latencyDetails: Option[LatencyDetails] = value.businesses.find(_.incomeSourceId.getOrElse("").equals(incomeSourceId)).flatMap(_.latencyDetails)
+        latencyDetails match {
+          case Some(x) =>
+            val currentTaxYearEnd = dateService.getCurrentTaxYearEnd(isEnabled(TimeMachineAddYear))
+            x match {
+              case LatencyDetails(_, _, _, ty2, _) if ty2.toInt < currentTaxYearEnd => Future.successful(None)
+              case LatencyDetails(_, ty1, ty1i, ty2, ty2i) =>
```
Comment (joesheridan) [inline app/services/BusinessReportingMethodService.scala:85]: can we be a bit more verbose? eg  ty => taxYear, ty1i => taxYearLatencyIndicator

File: app/services/BusinessReportingMethodService.scala (hunk 11 above)
Comment (joesheridan) [inline app/services/BusinessReportingMethodService.scala:60]: ty => taxYear pls 😄

File: app/services/BusinessReportingMethodService.scala [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+  }
+
+  private def annualQuarterlyToBoolean(method: Option[String]): Option[Boolean] = method match {
+    case Some("A") => Some(true)
+    case Some("Q") => Some(false)
+    case _ => None
+
+  }
+
+  def updateIncomeSourceTaxYearSpecific(nino: String, incomeSourceId: String, formData: Map[String, String])(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[Option[UpdateIncomeSourceResponse]] = {
```
Comment (joesheridan) [inline app/services/BusinessReportingMethodService.scala:117]: should be in 1566 service

File: app/models/calculationList/CalculationListModel.scala [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+package models.calculationList
+
+import play.api.libs.json.{Format, Json}
+
+sealed trait CalculationListResponseModel
+
+case class CalculationListModel(calculationId: String,
+                                calculationTimestamp: String,
+                                calculationType: String,
+                                crystallised: Option[Boolean]) extends CalculationListResponseModel
```
Comment (joesheridan) [inline app/models/calculationList/CalculationListModel.scala:26]: I presume None here maps to false? If so, can we flatten this to just Boolean?

Comment (EdwardDennis) [reply app/models/calculationList/CalculationListModel.scala:26]: It throws an ISE at the moment
This is an optional field in both calc APIs

File: app/controllers/incomeSources/add/BusinessReportingMethodController.scala (hunk 7 above)
Comment (SiddharthAtos) [reply app/controllers/incomeSources/add/BusinessReportingMethodController.scala:114]: this is to check false errors with form when the taxYear 1 is crystallised, where we don't show taxYear 1 reporting method radio option.  It evaluates as an error

File: app/models/itsaStatus/ITSAStatusResponse.scala (hunk 9 above)
Comment (SiddharthAtos) [reply app/models/itsaStatus/ITSAStatusResponse.scala:24]: Its according to schema

File: app/services/CalculationListService.scala [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  def getCalculationList(nino: Nino, taxYearRange: String)
+                        (implicit headerCarrier: HeaderCarrier): Future[CalculationListResponseModel] = {
+    Logger("application").debug(s"[CalculationService][getLatestCalculation] - " +
+      s"Requesting calculation list (1896) data from the backend with nino / taxYearRange: ${nino.value} - $taxYearRange")
+    incomeTaxViewChangeConnector.getCalculationList(nino, taxYearRange)
+  }
+
+  def isTaxYearCrystallised(taxYear: Int)(implicit user: MtdItUser[_], hc: HeaderCarrier, ec: ExecutionContext): Future[Option[Boolean]] = {
```
Comment (joesheridan) [inline app/services/CalculationListService.scala:47]: 👍

**Review by joesheridan — APPROVED** [review]

**Review by joesheridan — CHANGES_REQUESTED** [review]
still issues with businessreportingmethodservice I think

**Review by joesheridan — APPROVED** [review]

**Pull request merged**